requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+
+This RFC proposes both a technical and product solution to this problem: Store `person_id` and `person_properties` on the main events table.
+
+However, doing so will require breaking changes to the product. These changes are likely to cause some issues for customers, but these have either a negligible impact (affecting around 0.03% of events) or can be mitigated through documentation and education.
+
+## Proposed solution
+
+From a technical standpoint: Store `person_id` and `person_properties` on the main events table.
+
+From a product standpoint storing person_id and person_properties on events requires making the following changes:
+1. After the first session we will no longer merge events prior to login with the main person’s events.
+2. Person properties become immutable: Changing properties applies only on new events and past data is not updated.
+
+### Customer Impact
+
+#### Benefits
+For both changes the benefits are as follows:
+Significant short term performance gains
+Ability to support the largest tiers of customer now and in the future
+Standard analysis such as events from paying users over time will stop yielding misleading results when people start / stop paying (as they do today)
+
+
+#### 1. After the first session we will no longer merge events prior to login with the main person’s events.
+
+![image](https://user-images.githubusercontent.com/85295485/151368075-a7ab9dc7-ee8d-4cad-a454-2524161973f1.png)
+
+How we'll map persons to events going forward:
+1. On initial visit, users device gets assigned an an (anonymous) distinct id `anonid1`. Posthog internally associates person_id `p1` to `anonid1`, which is saved together with their events
+2. When after signing up `posthog.identify('userid')` is called, we associate the same person_id `p1` with the new distinct id `userid`
+3. If the user later visits the site with a new device, a new anonymous distinct id `anonid2` is created. Posthog internally associates these events with a new person id `p2` as no mapping for that id exists.
+4. If after signing in `posthos.identify('userid')` is called, we associate events going forward with person_id `p1`
+
+As a result, events done between steps 1, 2 and 4 will be tracked as from person `p1` but events between steps 3 and 4 will be under person`p2`. As a result no events are lost but some will be considered as coming from a different (anonymous) person. In the above image this corresponds to the green period.
+
+This affects use cases where users access their product (Logged out) on multiple devices regularly. However, based on our analysis the impact is only on ~0.03% events, some marketing use-cases are more significantly affected.
```

Comment:
Could you elaborate on the marketing use-cases that would be affected? Most marketing websites don't `posthog.identify` users at all and primarily rely on cookies. I guess you're maybe thinking about tracking re-activation of users?

Reply:
A sample one is this story:

1. User signs up
2. Some time later user becomes inactive
3. App sends user an email
4. User opens link in a re-activation/marketing email on a new device, signed out
5. User signs in and becomes active

If you wanted to track users who opened the link due to this change you will need some custom code on your end.

cc @marcushyett-ph you might have a few others.

Reply:
Yep its primarily around re-activation of existing users on different devices (so the size of the use-case is generally very small). By marketing I was referring primarily to scenarios where the logged out experience (usually marketing site) is part of the end-to-end funnel.